url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/Signaler.java

```diff
@@ -0,0 +1,172 @@
+/*
+    Copyright (c) 2010-2011 250bpm s.r.o.
+    Copyright (c) 2010-2011 Other contributors as noted in the AUTHORS file
+
+    This file is part of 0MQ.
+
+    0MQ is free software; you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    0MQ is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+package zmq;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.SelectableChannel;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.nio.channels.Pipe;
+import java.util.concurrent.atomic.AtomicInteger;
+
+//  This is a cross-platform equivalent to signal_fd. However, as opposed
+//  to signal_fd there can be at most one signal in the signaler at any
+//  given moment. Attempt to send a signal before receiving the previous
+//  one will result in undefined behaviour.
+
+public class Signaler {
+    //  Underlying write & read file descriptor.
+    private Pipe.SinkChannel w;
+    private Pipe.SourceChannel r;
+    private Selector selector;
+    private ByteBuffer rdummy;
+ 
+    // Selector.selectNow at every sending message doesn't show enough performance
+    private final AtomicInteger wcursor = new AtomicInteger(0);
+    private int rcursor = 0;
+    
+    public Signaler() {
+        //  Create the socketpair for signaling.
+        make_fdpair ();
+
+        //  Set both fds to non-blocking mode.
+        try {
+            Utils.unblock_socket (w);
+            Utils.unblock_socket (r);
+        } catch (IOException e) {
+            throw new ZError.IOException(e);
+        }
+        
+        try {
+            selector = Selector.open();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource that might leak to a class field, causing a resource leak that slows down or crashes your system. For example, when SelectableChannel.register throws an exception, the *Signaler* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *selector* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).